store/store_k8s.go

```diff
@@ -0,0 +1,257 @@
+		} else {
+			sanitizedString = k8sName
+		}
+	}
+	return sanitizedString
+}
+
+var (
+	// total wait time: 16.25 seconds
+	waitBackoff = wait.Backoff{
+		Duration: waitDuration, // the base duration
+		Factor:   waitFactor,   // Duration is multiplied by factor each iteration
+		Steps:    waitSteps,    // Exit with error after this many steps
+	}
+	etcdErrorsToRetry = []error{rpctypes.ErrLeaderChanged}
+)
+
+type k8sStore struct {
+	cm configmap.ConfigMap
+}
+
+// NewK8sStore returns a Store implementation which uses
+// k8s configmaps to store data.
+func NewK8sStore(clusterID string) (Store, configmap.ConfigMap, error) {
+	k8sStore, cm, err := NewK8sStoreWithParams(
+		configmap.GetName(confgMapPrefix, clusterID),
+		configmap.DefaultK8sLockTimeout,
+		configmap.DefaultK8sLockAttempts * time.Second,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	return k8sStore, cm, nil
+}
+
+// NewK8sStoreWithParams returns a Store implementation which uses
+// k8s configmaps to store data. ConfigMap properties can be customized.
+func NewK8sStoreWithParams(
+	name string,
+	lockTryDuration time.Duration,
+	lockTimeout time.Duration,
+) (Store, configmap.ConfigMap, error) {
+	lockAttempts := uint((lockTryDuration / time.Second))
+	cm, err := configmap.New(
+		name,
+		nil,
+		lockTimeout,
+		lockAttempts,
+		0,
+		0,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	return &k8sStore{cm}, cm, nil
+}
+
+func (k8s *k8sStore) Lock(owner string) (*StoreLock, error) {
+	if err := k8s.cm.Lock(owner); err != nil {
+		return nil, err
+	}
+	return &StoreLock{owner: owner}, nil
+}
+
+func (k8s *k8sStore) LockWithKey(owner, key string) (*StoreLock, error) {
+	if err := k8s.cm.LockWithKey(owner, key); err != nil {
+		return nil, err
+	}
+	return &StoreLock{Key: key, owner: owner, lockedWithKey: true}, nil
+}
+
+func (k8s *k8sStore) Unlock(storeLock *StoreLock) error {
+	if storeLock.lockedWithKey {
+		return k8s.cm.UnlockWithKey(storeLock.Key)
+	}
+	return k8s.cm.Unlock()
+}
+
+func (k8s *k8sStore) IsKeyLocked(key string) (bool, string, error) {
+	return k8s.cm.IsKeyLocked(key)
+}
+
+func (k8s *k8sStore) CreateKey(key string, value []byte) error {
+	sanitizedKey := GetSanitizedK8sName(key)
+	err := k8s.cm.LockWithKey(string(value), sanitizedKey)
+	if err != nil {
+		logrus.Errorf("unable to lock with key %v", key)
+		return err
+	}
+	defer func() {
+		err := k8s.cm.UnlockWithKey(sanitizedKey)
+		if err != nil {
+			logrus.Infof("unable to unlock with key %v", key)
+		}
+	}()
+
+	data, err := k8s.cm.Get()
+	if err != nil {
+		return err
+	}
+
+	if _, ok := data[key]; ok {
+		return &StoreKeyExists{
+			Key:     key,
+			Message: "Use PutKey API",
+		}
+	}
+
+	if data == nil {
+		data = make(map[string]string)
+	}
+	data[key] = string(value)
+	return k8s.patchWithRetries(data)
+}
+
+func (k8s *k8sStore) PutKey(key string, value []byte) error {
+	data, err := k8s.cm.Get()
+	if err != nil {
+		return err
+	}
+
+	data[key] = string(value)
```

Comment:
can we safely assume that the value can always be converted to string without requiring any encoding e.g. base64? If yes, should we just change the interface to accept string instead of []byte to make that clear?

Reply:
It appears so. May be @adityadani can add more context to this.

Reply:
https://portworx.atlassian.net/browse/PWX-30066